src/OpenLoco/src/Windows/ObjectSelectionWindow.cpp

```diff
@@ -99,21 +159,112 @@ namespace OpenLoco::Ui::Windows::ObjectSelectionWindow
     }
 
     // 0x004731EE
-    static void sub_4731EE(Window* self, ObjectType eax)
+    static void repositionTargetTab(Window* self, ObjectType targetTab)
     {
-        registers regs;
-        regs.eax = static_cast<uint32_t>(eax);
-        regs.esi = X86Pointer(self);
-        call(0x004731EE, regs);
+        self->currentTab = enumValue(targetTab);
+        for (auto i = 0U; i < std::size(_tabInformation); i++)
+        {
+            // Ended up in a position without info? Reassign positions first.
+            if (_tabInformation[i].index == 0xFF)
+            {
+                self->var_856 |= (1 << 0);
+                assignTabPositions(self);
+                continue;
+            }
+
+            if (_tabInformation[i].index == enumValue(targetTab))
+            {
+                // Found current tab, and its in bottom row? No change required
+                if (_tabInformation[i].row == 0)
+                    return;
+                // Otherwise, we'll rotate the tabs around, such that this one is in the bottom row
+                else
+                    return rotateTabs(i);
+            }
+        }
+    }
+
+    // 0x00473154
+    static void assignTabPositions(Window* self)
+    {
+        uint8_t currentRow = 0;
+        uint8_t currentPos = 0;
+        uint8_t rowCapacity = 19;
+        uint8_t tabPos = 0;
+
+        for (int8_t currentType = ObjectManager::maxObjectTypes - 1; currentType >= 0; currentType--)
+        {
+            const ObjectTabFlags tabFlags = _tabDisplayInfo[currentType].flags;
+
+            if ((tabFlags & ObjectTabFlags::alwaysHidden) != ObjectTabFlags::none)
+                continue;
+
+            // Skip all types that don't have any objects
+            if (_tabObjectCounts[currentType] == 0)
+                continue;
+
+            // Skip certain object types that only have one entry in game
+            if ((tabFlags & ObjectTabFlags::showEvenIfSingular) == ObjectTabFlags::none && _tabObjectCounts[currentType] == 1)
+                continue;
+
+            // Hide advanced object types as needed
+            if ((self->var_856 & (1 << 0)) == 0 && (tabFlags & ObjectTabFlags::advanced) != ObjectTabFlags::none)
+                continue;
+
+            if (isEditorMode() && (tabFlags & ObjectTabFlags::hideInEditor) != ObjectTabFlags::none)
+                continue;
+
+            if ((tabFlags & ObjectTabFlags::hideInGame) != ObjectTabFlags::none)
+                continue;
+
+            // Assign tab position
+            _tabInformation[tabPos].index = static_cast<uint8_t>(currentType);
+            _tabInformation[tabPos].row = currentRow;
+            tabPos++;
+
+            // Distribute tabs over two rows -- ensure there's capacity left in current row
+            currentPos++;
+            if (currentPos >= rowCapacity)
+            {
+                currentPos = 0;
+                rowCapacity--;
```

Comment:
i think i'd prefer just two constants for capacity. 19 for row 0 and 18 for second row.